### PR TITLE
Variable ac-zeta

### PR DESCRIPTION
**Pull request by sambitmishra98**
This PR is the first step to varying ac-zeta across P-multigrid levels. I have compared csv files of `[soln-plugin-pseudostats]` for the develop version and the PR, to check if the PR works. Setting ac-zeta this way did not seem to affect the performance of the inc-cylinder-2d case from PyFR-test-cases. Let me know if we need to test its performance difference with the develop version more.

Also, I was not sure where to initialise ac-zeta, so I shifted it from `[constants]` to `[solver]`.

With this PR, we can now look into the effect of changing ac-zeta across pseudo-iterations, multi-P levels, and stages.

### CODE REVIEW

File: pyfr/solvers/aceuler/system.py [hunk 1]
```diff
@@ -11,3 +11,25 @@ class ACEulerSystem(BaseAdvectionSystem):
...
+            b.prepare(t)
+
+        for b in binders: 
+            b(t=t, ac_zeta=self._ac_zeta)
+
+    @property
+    def ac_zeta(self):
+        return self._ac_zeta
+
+    @ac_zeta.setter
```
Comment (FreddieWitherden) [inline pyfr/solvers/aceuler/system.py:33]: Why are these needed?

File: pyfr/solvers/aceuler/system.py [hunk 2]
```diff
@@ -11,3 +11,25 @@ class ACEulerSystem(BaseAdvectionSystem):
     intinterscls = ACEulerIntInters
     mpiinterscls = ACEulerMPIInters
     bbcinterscls = ACEulerBaseBCInters
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        
+        self._ac_zeta = self.cfg.getfloat('solver', 'ac-zeta')
+
+    def _prepare_kernels(self, t, uinbank, foutbank):
```
Comment (FreddieWitherden) [inline pyfr/solvers/aceuler/system.py:20]: Too much code duplication here.

File: pyfr/solvers/acnavstokes/elements.py [hunk 3]
```diff
@@ -37,24 +37,28 @@ def set_backend(self, *args, **kwargs):
             self.kernels['tdisf_curved'] = lambda uin: self._be.kernel(
                 'tflux', tplargs=tplargs, dims=[self.nupts, r[c]],
                 u=s(self.scal_upts[uin], c), f=s(self._vect_upts, c),
+                extrns={'ac_zeta': 'scalar fpdtype_t'},
```
Comment (FreddieWitherden) [inline pyfr/solvers/acnavstokes/elements.py:40]: See if we can factor this out.  For example define:
```python
def kernel(*args, **kwargs):
    return lambda: self._be.kernel(*args, **kwargs, tplargs=tplargs,
                                   externs={'ac_zeta': 'scalar fpdtype_t'})
```

File: pyfr/solvers/aceuler/inters.py [hunk 4]
```diff
@@ -45,10 +49,10 @@ def __init__(self, *args, **kwargs):
         tplargs = dict(ndims=self.ndims, nvars=self.nvars, rsolver=rsolver,
                        c=self.c, bctype=self.type)
 
+        self._set_external('ac_zeta', 'scalar fpdtype_t')
         self.kernels['comm_flux'] = lambda: self._be.kernel(
             'bccflux', tplargs=tplargs, dims=[self.ninterfpts],
             extrns=self._external_args, ul=self._scal_lhs, nl=self._pnorm_lhs,
-            **self._external_vals
```
Comment (FreddieWitherden) [inline pyfr/solvers/aceuler/inters.py:51]: Why are we removing this?

**Comment by FreddieWitherden** [comment]
What kind of improvement does this have the potential to give for a real-world test case like a TGV or SD7003?

**Comment by sambitmishra98** [comment]
I just tested it for a TGV case. I added a few lines to make `ac-zeta` vary across multi-P levels by a factor, just like `pseudo-dt-fact`.

```
mcfg.set('solver', 'ac-zeta', self.zeta*self.zetaf**(order - l))
```
|ac-zeta-fact|total pseudo-steps|total-time|
| -- | -- | -- |
| 1.0 (develop) | 4062 | 12m 56s |
| 1.0 (PR) | 4062 | 13m 00s |
| 1.5 (PR) | 3982 | 12m 48s |
| 2.0 (PR) | 3996 | 12m 57s |

Last time I tried, I had got speedup by changing with pseudo-iterations. Let me do that and get the speed-up before proceeding any further.

File: pyfr/solvers/aceuler/elements.py [hunk 5]
```diff
@@ -62,24 +62,28 @@ def set_backend(self, *args, **kwargs):
             self.kernels['tdisf_curved'] = lambda uin: self._be.kernel(
                 'tflux', tplargs=tplargs, dims=[self.nupts, r[c]],
                 u=s(self.scal_upts[uin], c), f=s(self._vect_upts, c),
+                extrns={'ac_zeta': 'scalar fpdtype_t'},
```
Comment (WillTrojak) [inline pyfr/solvers/aceuler/elements.py:65]: Could you set this as a variable eg `exargs` and then pass that to each kernel? Should reduce the chance of any errors.

File: pyfr/solvers/acnavstokes/system.py [hunk 6]
```diff
@@ -12,3 +12,25 @@ class ACNavierStokesSystem(BaseAdvectionDiffusionSystem):
     intinterscls = ACNavierStokesIntInters
     mpiinterscls = ACNavierStokesMPIInters
     bbcinterscls = ACNavierStokesBaseBCInters
+
```
Comment (WillTrojak) [inline pyfr/solvers/acnavstokes/system.py:15]: Maybe this could be a mixin? `ExternalSystem` or something. You will probably need to come up with a nice syntax for saying what externals are needed though.

**Review by WillTrojak — COMMENTED** [review]
Just a few minor changes; the main thing that needs thought is the repetition in the system class.

**Comment by FreddieWitherden** [comment]
Did this turn out to be beneficial in the end?

**Comment by sambitmishra98** [comment]
No, nothing conclusive.
